### PR TITLE
added new transforms to support speaker/asr applications, upgraded transform numerical accuracy

### PR DESCRIPTION
**Pull request by ranchlai**
The new transforms are as follows:
'Noisify',
'Reverberate',
'RandomChoice',
'RandomApply',
See the docstring for more details.
Numerical accuracy compare to librosa is up to 1e-7 for stft and melspectrogram transform
see test/unit_test for more details.

### CODE REVIEW

**Review by ZeyuChen — COMMENTED** [review]
It would be better add example to confirm these Transform function is actually needed. Or please add more background information or opensource project to indicate which asr or speaker classical model need them.

**Comment by ZeyuChen** [comment]
Since @KPatr1ck wrote a exmaples of speaker also with RIR, what's the relationship and what's your plan to unify these transform?

File: PaddleAudio/paddleaudio/transforms.py [hunk 1]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
...
+    def __repr__(self):
+        format_string = self.__class__.__name__ + '('
+        for t in self.transforms:
+            format_string += '\n'
+            format_string += '    {0}'.format(t)
+        format_string += f'\n)'
+        return format_string
+
+
+class Noisify:
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py]: 如果是输入输出都是Tensor，建议统一继承自nn.Layer，然后用forward来完成计算逻辑。

File: PaddleAudio/paddleaudio/transforms.py [hunk 2]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
     def __repr__(self, ):
         return (self.__class__.__name__ +
                 f'(min_mu={self.min_mu}, max_mu={self.max_mu})')
+
+
+class RIRSource(nn.Layer):
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py]: 从文件读取RIR音频，从逻辑上更适合封装成dataset。
参考竞品，加载的逻辑自下到上应该是：dataset -> dataloader -> forward(实际操作者)
可以考虑下是否需将XXSource设计为dataset，让dataloader去作为”source“供迭代的方式更加合适。

File: PaddleAudio/paddleaudio/transforms.py [hunk 3]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
...
+            file = random.choice(self.rir_files)
+        else:
+            i = self.idx % self.n_files
+            file = self.rir_files[i]
+            self.idx += 1
+            if self.idx >= self.n_files:
+                self.idx = 0
+
+        rir, _ = paddleaudio.load(file, sr=self.sample_rate, mono=True)
+        rir_weight = paddle.to_tensor(rir[None, None, ::-1])
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py]: 如果按照上述用dataloader做，这里从loader取出就是tensor，不需要手动转。

File: PaddleAudio/paddleaudio/transforms.py [hunk 4]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
...
+        weight = self.rir_reader()  #get next weight
+        pad_len = [
+            weight.shape[-1] // 2 - 1, weight.shape[-1] - weight.shape[-1] // 2
+        ]
+        out = paddle.nn.functional.conv1d(x.unsqueeze(1),
+                                          weight,
+                                          padding=pad_len)
+        return out[:, 0, :]
+
+    def __repr__(self):
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py:922]: 无效信息打印。

File: PaddleAudio/paddleaudio/transforms.py [hunk 5]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
...
+        return (self.__class__.__name__)
+
+
+class RandomApply():
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py:926]: RandomApply和RandomChoice都是提供所有augment选项然后概率性地选择执行，放transforms里感觉有点怪。
或者可以封装成一个更大的augment类，提供概率信息和transform list，让这个大类在内部执行概率性选择和执行augment。

File: PaddleAudio/paddleaudio/transforms.py [hunk 6]
```diff
@@ -840,3 +850,370 @@ def forward(self, x: Tensor) -> Tensor:
...
+        return x
+
+    def __repr__(self):
+        return (
+            self.__class__.__name__ +
+            f'(random={self.random}, snr_high={self.snr_high}, snr_low={self.snr_low})'
+        )
+
+
+class AudioSource:
```
Comment (KPatr1ck) [inline PaddleAudio/paddleaudio/transforms.py]: 同上，files -> array -> tensor可以考虑封装成dataset

**Review by KPatr1ck — COMMENTED** [review]
对一些类的设计提了点建议～

File: PaddleAudio/paddleaudio/transforms.py (hunk 3 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: transform这里的想法是与dataloader剥离，并且不依赖于numpy。不过目前paddleaudio.load还是基于numpy。backend也将不依赖numpy

File: PaddleAudio/paddleaudio/transforms.py (hunk 1 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: 已改过来。

File: PaddleAudio/paddleaudio/transforms.py (hunk 6 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: 同上，transform是不希望与dataset关联。但是由于需要有信号源，所以定了这个类。AudioSource打算改名为NoiseSource。另外，后续会新增一些Source，比如WhiteNoiseSource等，是通过随机算法产生的。RIRSource也不一定需要从文件中读取。本质上这些source不是dataset。 dataset通常指用于学习的数据集，但这些source更偏信号处理。

File: PaddleAudio/paddleaudio/transforms.py (hunk 5 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py:926]: RandomApply和RandomChoice是transform中的主要工具函数。core library的设计我觉得不应该加入太多aumegntation的概念，比如Masking定义为一个transform，而不是augmentation。关于RandomApply等还可以参考vision里面的一些transform设计理念。
另, RandomApply和randomChoice两者的概率逻辑不同，前者每个tansform都是独立，后者是n选一。

File: PaddleAudio/paddleaudio/transforms.py (hunk 2 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: transform是不希望与dataset关联。但是由于需要有信号源，所以定了这个类。AudioSource打算改名为NoiseSource。另外，后续会新增一些Source，比如WhiteNoiseSource等，是通过随机算法产生的。RIRSource也不一定需要从文件中读取。本质上这些source不是dataset。 dataset通常指用于学习的数据集，但这些source更偏信号处理。

File: PaddleAudio/paddleaudio/transforms.py (hunk 4 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py:922]: 已增加了source信息显示。

File: PaddleAudio/paddleaudio/transforms.py (hunk 3 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: 未来backend将不用依赖于numpy，过度依赖于numpy不利于进行端到端delopyment

File: PaddleAudio/paddleaudio/transforms.py (hunk 2 above)
Comment (ranchlai) [reply PaddleAudio/paddleaudio/transforms.py]: 另一个不希望与dataloader关联的原因是 dataloader是多进程进行tensor处理，而paddle的tensor不能分别指定设备，使无法进行CPU/GPU混合计算（CPU下多进程，GPU下单进程）

File: PaddleAudio/paddleaudio/transforms.py (hunk 3 above)
Comment (ZeyuChen) [reply PaddleAudio/paddleaudio/transforms.py]: 如果都跟文件打交道了，作为transform操作还是有点奇怪。

File: PaddleAudio/paddleaudio/transforms.py (hunk 2 above)
Comment (ZeyuChen) [reply PaddleAudio/paddleaudio/transforms.py]: 两位建议横向看下竞品的逻辑经过对比后再决定模块的位置。

Comment (ZeyuChen) [reply PaddleAudio/paddleaudio/transforms.py]: 这个PR的合入要谨慎，没有充分的样例呈现和使用说明的话不能贸然合入。

**Review by ZeyuChen — COMMENTED** [review]
改PR需要经过进一步讨论与其他产品使用方式对比后再合入。

**Comment by ranchlai** [comment]
will open another PR
